### PR TITLE
Video prediction

### PR DESCRIPTION
**Pull request by danbider**
* refactored image plotting in fiftyone, now called `visualize_frame_preds.py`
* some fixes in `plotting_utils.py`
* a hydra script for predicting new vids `predict_new_vids.py`
* a hydra script for rendering vids based on csv predictions from multiple models `render_labeled_vids.py`
* style + explanation changes to `README.md`
* fixed some paths in tests
* making sure that the following works: full training -> prediction on train/test/val -> fiftyone -> prediction on new test vids

Open Issues:
* using actual keypoint names in `plotting_utils.py` - should we use a list of bp names to initialize the models themselves, instead of specifying an integer like `num_keypoints=17`?
* deciding on simple conventions like -- how should we name the saved csv, where do we want the labeled videos to be saved?
* extending the `render_labeled_vids.py` script to handle multiple videos or even multiple folders. that depends on the above point.

### CODE REVIEW

File: pose_est_nets/utils/plotting_utils.py [hunk 1]
```diff
@@ -39,7 +96,9 @@ def saveNumericalPredictions(model, datamod, threshold):
     full_dl = datamod.full_dataloader()
     test_dl = datamod.test_dataloader()
     final_gt_keypoints = np.empty(shape=(len(test_dl), model.num_keypoints, 2))
-    final_imgs = np.empty(shape=(len(test_dl), 406, 396, 1))
+    final_imgs = np.empty(
+        shape=(len(test_dl), 406, 396, 1)
```
Comment (themattinthehatt) [inline pose_est_nets/utils/plotting_utils.py:100]: can we pull these numbers from the model somehow instead of hard-coding them?

**Review by themattinthehatt — CHANGES_REQUESTED** [review]

File: scripts/predict_new_vids.py [hunk 2]
```diff
@@ -34,12 +34,11 @@ def make_predictions(cfg: DictConfig):
...
         -1
     ]  # the output of glob.glob is a list with all files that match the search.
-
     predict_videos(
-        cfg.eval.path_to_test_videos,
-        model_ckpt_path,
-        model_config_path,
-        None,
+        video_path=cfg.eval.path_to_test_videos,
+        model_file=model_ckpt_path,
```
[the remark this answers is not in the document]
Comment (danbider) [reply scripts/predict_new_vids.py]: it's an outdated file, already called `ckpt_file` and `model_cfg`

File: scripts/train_hydra.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+from genericpath import isdir
```
[the remark this answers is not in the document]
Comment (danbider) [reply scripts/train_hydra.py:1]: deleted

File: pose_est_nets/utils/plotting_utils.py [hunk 4]
```diff
@@ -260,7 +262,7 @@ def predict_videos(
     else:
         raise ValueError("cfg_file must be str or DictConfig, not %s!" % type(cfg_file))
 
-    model = load_model_from_checkpoint(cfg=cfg, ckpt_file=ckpt_file)
+    model = load_model_from_checkpoint(cfg=cfg, ckpt_file=ckpt_file, eval=True)
 
     model.to(device_pt)
     model.eval()
```
[the remark this answers is not in the document]
Comment (danbider) [reply pose_est_nets/utils/plotting_utils.py:285]: done, good idea

File: pose_est_nets/utils/plotting_utils.py [hunk 5]
```diff
@@ -308,8 +310,11 @@ def predict_videos(
         n_frames_ = count_frames(video_file)  # total frames in video
         n_frames = 0  # total frames processed
         keypoints_np = np.zeros((n_frames_, model.num_targets))
+        confidence_np = np.zeros((n_frames_, model.num_targets // 2))
```
[the remark this answers is not in the document]
Comment (danbider) [reply pose_est_nets/utils/plotting_utils.py:330]: done

File: scripts/render_labeled_vids.py [hunk 6]
```diff
@@ -42,43 +42,54 @@ def render_labeled_videos(cfg: DictConfig):
     # there may be multiple video paths in the cfg eval
     # certainly there may be multiple models.
     # loop over both
-    video_dir = verify_absolute_path(cfg.eval.path_to_test_videos)
+    video_dir = verify_absolute_path(
+        cfg.eval.path_to_test_videos[0]
+    )  # TODO: just for now assuming there's one dir, but can have multiple dirs.
     video = get_videos_in_dir(video_dir)
     assert (
```
[the remark this answers is not in the document]
Comment (danbider) [reply scripts/render_labeled_vids.py:55]: done

File: pose_est_nets/utils/plotting_utils.py (hunk 1 above)
Comment (danbider) [reply pose_est_nets/utils/plotting_utils.py:100]: this is an unused and outdated function, i'd either remove it or ask @nickgreenspan to fix it?
I'd save stuff as a `.csv` that looks exactly like the labels csv.
